### ISSUE TITLE
expose Compiler.compile_data after compile

### ISSUE THREAD

**Issue by sanderr** [inmanta/inmanta#2067]
Currently a successful compile only exposes the types and the root namespace. It might be useful to expose `inmanta.compiler.Compiler.compile_data` as well. Currently (#2048) this only contains information about errors occurred during the compile run, but as the JSON export API grows, so will the compile_data object. It might be useful to be able to access the same data internally (from plugins, from data trace application, ...).
A failed compile currently attaches the root namespace to exceptions iff the data trace is enabled (`inmanta.compiler.Compiler._handle_exception_datatrace` and `inmanta.ast.CompilerException.attach_compile_info`). Compile data should be attached as well or the compiler instance should be attached instead, exposing both the namespace and the compile_data instance.
This should be documented under the compiler internal API section in the documentation.

**Comment by wouterdb** [comment]
@sander: expose Compiler.compile_data where?

**Comment by sanderr** [comment]
To the internal API

**Comment by sanderr** [comment]
More specific: return it from `inmanta.compiler.do_compile()` in some way.

**Comment by wouterdb** [comment]
Agreed. Perhaps we have to introduce a new method that will deprecate do_compile, that provides a more structured output (e.g. an output object, that collect the various forms of output, that can be extended with additional fields in the future.)

**Comment by sanderr** [comment]
That sounds like a good solution. I think `CompileData` might be a candidate for such an output object, at least if we want to be able to export this output in some way as well, which I think we do.

**Comment by wouterdb** [comment]
We may want to export all the output, but we will  need fine-grained control over what goes where/when.

(e.g. model export could be integrated here as well, but normal compiler output is not expected to contain the full model)

**Comment by sanderr** [comment]
Can the suggestion and discussion labels be removed?
